### ISSUE TITLE
S2583: False-positive when Monitor.Wait is used

### ISSUE THREAD

**Issue by shadow-cs** [SonarSource/sonar-csharp#947]
### Description

SonarLint detects following code as a violation of [S2583](https://rules.sonarsource.com/csharp/RSPEC-2583) due to a duplicate check with the same expression, but since `Monitor.Wait` is used in the middle the `value` may indeed change (by other thread).

Marking the `value` as `volatile` doesn't help either (but it should have no effect on the result since the lock makes sure all variables are read correctly).

### Repro steps

```c#
private bool value = false;
private readonly object mutex = new object();

private void Foo()
{
	lock (mutex)
	{
		if (value)
			return;
		Monitor.Wait(mutex);
		if (value)
			return;
	}
}
```

### Expected behavior

No issue detected.

### Actual behavior

The last check for `value` is detected as an issue.

### Known workarounds

Suppress the warning manually.

### Related information

* SonarC# 3.7.0.2645
* Visual Studio 15.4.4

**Comment by valhristov** [comment]
Thanks for submitting this issue. SonarC# analyzes methods one by one and is unable to detect multithreaded execution. For now, without knowing your actual scenario, the best solution would be to mark the raised issues as False Positive, perhaps in the future we would be able to detect problems like this and reduce the noise.

**Comment by shadow-cs** [comment]
Understood, I know that multi-threaded execution analysis is not handled by SonarC# (understandably).
But I'd assume that certain use cases (where common system methods are used) could bypass some checks. In my scenario `value` is a field and `Monitor.Wait` could be treated as calling another instance method (since it it possible the value might change during its execution).

I know that not all use cases could possibly be handled (there are so many synchronization primitives that could provide similar behavior, some of them not being part of CLR) but the most common ones (especially using the `Monitor`) could be handled. Just an opinion ;-)

**Comment by DuAell** [comment]
Had the same issue. SonarC# should detect anything when it involves a member or property outside of the analysed method scope.
